config/initializers/rack_attack.rb

```diff
@@ -81,25 +84,32 @@ def self.protected_route?(protected_actions, path, method)
   # throttle logins for another user and force their login requests to be
   # denied, but that's not very common and shouldn't happen to you. (Knock
   # on wood!)
-  throttle("logins/handler", limit: REQUEST_LIMIT, period: LIMIT_PERIOD) do |req|
-    if req.path == "/session" && req.post?
-      # return the handler if present, nil otherwise
-      req.params['session']['who'].presence if req.params['session']
+  protected_sessions_action = [{ controller: "sessions", action: "create" }]
+
+  throttle("logins/handle", limit: REQUEST_LIMIT, period: LIMIT_PERIOD) do |req|
+    if protected_route?(protected_sessions_action, req.path, req.request_method) && req.params['session']
+      User.normalize_email(req.params['session']['who'].presence)
     end
   end
 
+  throttle("api_key/basic_auth", limit: REQUEST_LIMIT, period: LIMIT_PERIOD) do |req|
+    req.env["HTTP_AUTHORIZATION"].presence if protected_route?(protected_api_key_action, req.path, req.request_method)
```

Comment:
One more thought.

I know you've previously mentioned a desire to keep the throttles in the one file, however consider how these changes above might be a little simpler if the throttle was placed closer to the code its trying to protect. 

In this case, the throttle would be inside the `User.authenticate` method. This has some  advantages:

- Less effort required to get the value of `who` to use as the throttle key as its already an argument to `User.authenticate`
- Any future controllers that make user of `User.authenticate` will already be protected against brute force attacks. This is a more secure default, as developers can forget to add a throttle to `rack_attack.rb` but the authentication is still protected.

Reply:
> The Authorization header value (req.env["HTTP_AUTHORIZATION"]) will be different for each password variation

Oh, yes. Updated.
> it uses an identical implementation to
> Rails' `authenticate_or_request_with_http_basic` method.
> req = ActionDispatch::Request.new(req.env)

`req` would be a fairly large object (`ObjectSpace.memsize_of(req) = 140`) and allocated on every request. I think to decode directly would have lesser overhead.

> who is currently used in 2 variations in User.authenticate

I forgot we had overriden `authenticate`, sorry. Using `normalize_email` in `authenticate` or in sign in rate limit is not as straight forward. As of now, existing users' email and handle with spaces and different casing are considered different users. I check our database and there is a good number of users who are depending on this. 
New users are validated for spaces. We don't have any plan of making handle case-sensitive, email casing is being discussed [here](https://github.com/rubygems/rubygems.org/issues/1763).

> how these changes above might be a little simpler if the throttle was placed closer to the code its trying to protect.

I agree issues I pointed above would be simpler if we throttle from model/controller. My preference is based on the rack-attack docs, where it says to use `config/initializers/rack_attack.rb (for rails app)`. How would rack-attack from any place besides initializer work? Are you suggesting we drop rack-attack and implement our own?

Reply:
Rack-attack would continue to be used:

```rb
def self.authenticate(who, password)
  user = find_by(email: who.downcase) || find_by(handle: who)
  throttle("user authentication", discriminator: user.id, limit: 10, period: 10.minutes)
  user if user&.authenticated?(password)
end
```

To implement `#throttle` we could borrow heavily from https://github.com/kickstarter/rack-attack/blob/master/lib/rack/attack/throttle.rb#L24-L53 and raise an error if the limit is exceeded.

```rb
def throttle(name, discriminator:, limit:, period:)
  # Calls Rack::Attack.cache.count, performs any instrumentation,
  # see Rack::Attack::Throttle#matched_by?
  ...
  raise TooManyAttempts if limit exceeded
end
```

Higher up the stack, this error would be translated to a 429 response, possibly using `Rack::Attack.throttled_response.call(request.env)` to be consistent in the 429 responses.

Reply:
(If anyone reading this is uncomfortable about using `CurrentAttributes`, I understand. Do consider if applied here it will improve reuse of `Rack::Attack::Throttle`).

Here's an alternative that makes better reuse of rack-attack's existing throttle code. 

- Introduce `Current.request` which stores the current request. [`CurrentAttributes`](https://api.rubyonrails.org/classes/ActiveSupport/CurrentAttributes.html) will come in handy for this.

- In `User.authenticate` reuse `Rack::Attack::Throttle`:

```rb
# app/models/user.rb
def self.authenticate(who, password)
  user = find_by(email: who.downcase) || find_by(handle: who)
  return unless user

  # limit and period are just example values
  throttle =
    Rack::Attack::Throttle.new("user authentication", limit: 10, period: 10.minutes) do |_req|
      # user.id as discriminator sidesteps issues with normalizing handles and emails.
      user.id
    end
  raise TooManyAttempts if throttle.matched_by?(Current.request)

  user if user.authenticated?(password)
end
```

This could be refined to something like:

```rb
# app/models/user.rb
def self.authenticate(who, password)
  user = find_by(email: who.downcase) || find_by(handle: who)
  return unless user

  # limit and period are just example values
  throttle("user authentication", limit: 10, period: 10.minutes) { user.id }

  user if user.authenticated?(password)
end

...

def self.throttle(name, options, &block)
  throttle = Rack::Attack::Throttle.new(name, options, &block)
  raise TooManyAttempts if throttle.matched_by?(Current.request)
end
```

I've a preference for this, mainly as it doesn't involve copying, pasting, and tweaking the bulk of `Rack::Attack::Throttle#matched_by?` which was the basis of the suggestion in my previous comment.

Reply:
I think it would be better if we take this up in a separate PR. Let me know you like to see any more changes beside it in this PR.

Reply:
Sounds good and looks great 👍